src/heap_roots.h

```diff
@@ -20,43 +20,60 @@
 
 namespace toit {
 
-class ObjectNotifier;
 class FinalizerNode;
+class Heap;
+class LivenessOracle;
+class ObjectNotifier;
+class ToitFinalizerNode;
 class VmFinalizerNode;
 
 typedef LinkedFifo<FinalizerNode> FinalizerNodeFifo;
 
 class FinalizerNode : public FinalizerNodeFifo::Element {
  public:
-  FinalizerNode(HeapObject* key, Object* lambda)
+  virtual ~FinalizerNode();
+  virtual bool has_key(HeapObject* value) = 0;
```

Comment:
If the `key_` field was in `FinalizerNode`, you could devirtualize some of this.